src/main/java/org/prebid/server/bidder/sonobi/SonobiBidder.java

```diff
@@ -1,20 +1,143 @@
 import com.iab.openrtb.request.Imp;
-import org.prebid.server.bidder.OpenrtbBidder;
+import com.iab.openrtb.response.Bid;
+import com.iab.openrtb.response.BidResponse;
+import com.iab.openrtb.response.SeatBid;
+import io.vertx.core.http.HttpMethod;
+import org.apache.commons.collections4.CollectionUtils;
+import org.prebid.server.bidder.Bidder;
+import org.prebid.server.bidder.model.BidderBid;
+import org.prebid.server.bidder.model.BidderError;
+import org.prebid.server.bidder.model.HttpCall;
+import org.prebid.server.bidder.model.HttpRequest;
+import org.prebid.server.bidder.model.Result;
+import org.prebid.server.exception.PreBidException;
+import org.prebid.server.json.DecodeException;
 import org.prebid.server.json.JacksonMapper;
+import org.prebid.server.proto.openrtb.ext.ExtPrebid;
 import org.prebid.server.proto.openrtb.ext.request.sonobi.ExtImpSonobi;
+import org.prebid.server.proto.openrtb.ext.response.BidType;
+import org.prebid.server.util.HttpUtil;
 
-public class SonobiBidder extends OpenrtbBidder<ExtImpSonobi> {
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+public class SonobiBidder implements Bidder<BidRequest> {
+
+    private static final TypeReference<ExtPrebid<?, ExtImpSonobi>> SONOBI_EXT_TYPE_REFERENCE =
+            new TypeReference<>() {
+            };
+
+    private final String endpointUrl;
+    private final JacksonMapper mapper;
 
     public SonobiBidder(String endpointUrl, JacksonMapper mapper) {
-        super(endpointUrl, RequestCreationStrategy.REQUEST_PER_IMP, ExtImpSonobi.class, mapper);
+        this.endpointUrl = HttpUtil.validateUrl(Objects.requireNonNull(endpointUrl));
+        this.mapper = Objects.requireNonNull(mapper);
     }
 
     @Override
-    protected Imp modifyImp(Imp imp, ExtImpSonobi impExt) {
-        return imp.toBuilder()
-                .tagid(impExt.getTagId())
+    public Result<List<HttpRequest<BidRequest>>> makeHttpRequests(BidRequest bidRequest) {
+        final List<HttpRequest<BidRequest>> requests = new ArrayList<>();
+        final List<BidderError> errors = new ArrayList<>();
+
+        for (Imp imp : bidRequest.getImp()) {
+            try {
+                final ExtImpSonobi extImpSonobi = parseImpExt(imp);
+                final Imp modifiedImp = modifyImp(imp, extImpSonobi.getTagId());
+                requests.add(makeRequest(bidRequest, modifiedImp));
+            } catch (PreBidException e) {
+                errors.add(BidderError.badInput(e.getMessage()));
+            }
+        }
+
+        return Result.of(requests, errors);
+    }
+
+    private ExtImpSonobi parseImpExt(Imp imp) throws PreBidException {
+        try {
+            return mapper.mapper().convertValue(imp.getExt(), SONOBI_EXT_TYPE_REFERENCE).getBidder();
+        } catch (IllegalArgumentException e) {
+            throw new PreBidException(e.getMessage(), e);
```

Comment:
Pass full exception has no sense since in makeHttpRequests you use only `e.getMessage()`